modules/cog-fdo-shell.c

```diff
@@ -81,19 +411,363 @@ cog_fdo_shell_class_init (CogFdoShellClass *klass)
     TRACE ("");
     CogShellClass *shell_class = COG_SHELL_CLASS (klass);
     shell_class->is_supported = cog_fdo_shell_is_supported;
+
+    wl_data.resize_window = resize_window;
+    wl_data.handle_key_event = handle_key_event;
 }
 
 static void
 cog_fdo_shell_class_finalize (CogFdoShellClass *klass)
 {
     TRACE ("");
+    /* TODO
+    g_assert (platform);
+
+    // free WPE view data
+    if (wpe_view_data.frame_callback != NULL)
+        wl_callback_destroy (wpe_view_data.frame_callback);
+    if (wpe_view_data.image != NULL) {
+        wpe_view_backend_exportable_fdo_egl_dispatch_release_exported_image (wpe_host_data.exportable,
+                                                                             wpe_view_data.image);
+    }
+    g_clear_pointer (&wpe_view_data.buffer, wl_buffer_destroy);
+
+    / * @FIXME: check why this segfaults
+    wpe_view_backend_destroy (wpe_view_data.backend);
+    * /
+
+    / * free WPE host data * /
+    / * @FIXME: check why this segfaults
+    wpe_view_backend_exportable_fdo_destroy (wpe_host_data.exportable);
+    * /
+*/
+    clear_input ();
+
+    destroy_window ();
+    clear_egl ();
+    clear_wayland ();
+}
+static void
+resize_window (void)
+{
+    int32_t pixel_width = win_data.width * wl_data.current_output.scale;
+    int32_t pixel_height = win_data.height * wl_data.current_output.scale;
+
+    if (win_data.egl_window)
+        wl_egl_window_resize (win_data.egl_window,
+			                  pixel_width,
+			                  pixel_height,
+			                  0, 0);
+
+    wpe_view_backend_dispatch_set_size (wpe_view_data.backend,
+                                        win_data.width,
+                                        win_data.height);
+    g_debug ("Resized EGL buffer to: (%u, %u) @%ix\n",
+            pixel_width, pixel_height, wl_data.current_output.scale);
 }
 
+
+static bool
+capture_app_key_bindings (uint32_t keysym,
+                          uint32_t unicode,
+                          uint32_t state,
+                          uint8_t modifiers)
+{
+    CogLauncher *launcher = cog_launcher_get_default ();
+    WebKitWebView *web_view = (WebKitWebView *)
```

Comment:
`cog_shell_get_view` a `CogView` ( what is defined witht `G_DEFINE_TYPE_WITH_PRIVATE (CogView, cog_view, WEBKIT_TYPE_WEB_VIEW`). I don't know the right syntax to get the actual `WebView`

Reply:
You can just use the GObject macro to cast to a `WebKitWebView`, that is `WEBKIT_WEB_VIEW`. It's the same as a direct cast but with type-checking.

Reply:
done